### PR TITLE
Use the decorator design pattern to correct the persons names

### PR DESCRIPTION
**Pull request by sambeck87**
- Interface

    - Create a class Nameable.
        - Implement a method called correct_name that will raise a NotImplementedError.

- Turn your Person class to Nameable

    - Make sure that your Person class inherits from Nameable
    - Make sure that this class has a method correct_name implemented. It should simply return the name attribute.

- Prepare base Decorator

    - Make sure that it inherits from Nameable.
    - In the constructor assign a nameable object from params to an instance variable.
    - Implement the correct_name method that returns the result of the correct_name method of the @nameable.

- Prepare CapitalizeDecorator and TrimmerDecorator

    - For the CapitalizeDecorator:
        - Create a class that inherits from the base Decorator class.
        - Implement a method correct_name that capitalizes the output of @nameable.correct_name.
    - For the TrimmerDecorator:
        - Create a class that inherits from the base Decorator class.
        - Implement a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name
+    raise NotImplementedError
```
Comment (GzimAsani) [inline nameable.rb]: - [ ] Good job adding NotImplementedError, but please add a message with it to make it easier for users to read 
```ruby
class Nameable
    def correct_name
        raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end
end
```

File: base_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+class BaseDecorator < Nameable
+  def initialize(nameable_object)
+    super()
+    @nameable_object = nameable_object
+  end
+
+  def correct_name
+    @nameable_object.correct_name
+  end
```
Comment (GzimAsani) [inline base_decorator.rb]: - [ ] Make nameable_object getter and setter by adding `attr_accessor :nameable` under the class definition

File: person.rb [hunk 3]
```diff
@@ -1,5 +1,8 @@
-class Person
+require './nameable'
```
Comment (GzimAsani) [inline person.rb]: - [ ] For files inside your project use `require_relative` also for better testing your app and connections add the `require_relative 'capitalize_decorator'`
`require_relative 'trimmer_decorator' ` as well

**Review by GzimAsani — CHANGES_REQUESTED** [review]
### Hello, @sambeck87                   :wave:   
 Good job so far,  done There are some issues that you still need to work on to go to the next project but you are almost there!
 
 ## Highlights
 * Good job with your project, it is almost done  :+1: 

 ## Required Changes 

- [ ] Check the comments below :arrow_double_down: 

 ## Optional suggestions
 _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding! clapclapclap
 
 Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.
 
 **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
 
 _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ShadyShawkat — APPROVED** [review]
## Hi @sambeck87, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**